pallets/stream-payment/src/lib.rs

```diff
@@ -0,0 +1,477 @@
+
+            // Emit event.
+            Pallet::<T>::deposit_event(Event::<T>::StreamClosed {
+                stream_id,
+                refunded: stream.deposit,
+            });
+
+            Ok(().into())
+        }
+
+        #[pallet::call_index(2)]
+        pub fn update_stream(
+            origin: OriginFor<T>,
+            stream_id: T::StreamId,
+        ) -> DispatchResultWithPostInfo {
+            // No problem with anyone updating any stream.
+            let _ = ensure_signed(origin)?;
+
+            let mut stream = Streams::<T>::get(stream_id).ok_or(Error::<T>::UnknownStreamId)?;
+            Self::perform_stream_payment(stream_id, &mut stream)?;
+            Streams::<T>::insert(stream_id, stream);
+
+            Ok(().into())
+        }
+
+        #[pallet::call_index(3)]
+        pub fn refill_stream(
+            origin: OriginFor<T>,
+            stream_id: T::StreamId,
+            increase: T::Balance,
+        ) -> DispatchResultWithPostInfo {
+            let origin = ensure_signed(origin)?;
+            let mut stream = Streams::<T>::get(stream_id).ok_or(Error::<T>::UnknownStreamId)?;
+
+            // Only source can refill stream
+            ensure!(origin == stream.source, Error::<T>::UnauthorizedOrigin);
+
+            // Source will not pay for drained stream retroactively, so we perform payment with
+            // what is left first.
+            Self::perform_stream_payment(stream_id, &mut stream)?;
+
+            // Increase deposit.
+            T::Assets::increase_deposit(stream.asset_id.clone(), &origin, increase)?;
+            stream.deposit = stream
+                .deposit
+                .checked_add(&increase)
+                .ok_or(Error::<T>::CurrencyOverflow)?;
+
+            // Emit event.
+            Pallet::<T>::deposit_event(Event::<T>::StreamRefilled {
+                stream_id,
+                increase,
+                new_deposit: stream.deposit,
+            });
+
+            // Update stream info in storage.
+            Streams::<T>::insert(stream_id, stream);
+
+            Ok(().into())
+        }
+
+        #[pallet::call_index(4)]
+        pub fn change_stream_rate(
+            origin: OriginFor<T>,
+            stream_id: T::StreamId,
+            new_rate_per_time_unit: T::Balance,
+        ) -> DispatchResultWithPostInfo {
+            let origin = ensure_signed(origin)?;
+            let mut stream = Streams::<T>::get(stream_id).ok_or(Error::<T>::UnknownStreamId)?;
+
+            // Only source or target can update the rate.
+            ensure!(
+                origin == stream.source || origin == stream.target,
+                Error::<T>::UnauthorizedOrigin
+            );
+
+            // Noop
+            if new_rate_per_time_unit == stream.rate_per_time_unit {
+                return Ok(().into());
+            }
+
+            // Ensure rate change is fair.
+            if origin == stream.source && new_rate_per_time_unit < stream.rate_per_time_unit {
```

Comment:
I think I would allow origin and target to both increase or decrease but I would make it such that:
- if your new rate is higher and you are the source, you need confirmation from the target
- if your new rate is smaller and you are the target, you need confirmation from the source

It probably is the same as what you have but at least two parties get involved in such case